Provide optimised codegen for small libc calls

Aaronepower opened the pull request:
resolves #542

sunfishcode commented on lib/frontend/src/frontend.rs:
```diff
@@ -614,8 +684,74 @@ impl<'a> FunctionBuilder<'a> {
+        }
+
+        next_power = next_power.next_power_of_two();
     }
```
Power-of-two multiples are interesting because when you multiply a number by a power of 2, it's effectively just shifting the number left, inserting zeros at the low end. The number of zeros inserted is equal to the number of trailing zeros in the power of 2.

So instead of iterating here, we can use `trailing_zeros` to find the number of trailing zeros in the size. Its greatest power-of-two divisor is then the power of two with that many trailing zeros, or a one left-shifted by that number.

sunfishcode commented on lib/frontend/src/frontend.rs:
```diff
@@ -571,10 +571,57 @@ impl<'a> FunctionBuilder<'a> {
+                             dest_align: u8,
+                             src_align: u8)
+    {
+        const THRESHOLD: u64 = 4;
```
Please add a comment mentioning that this size is just a guess, not based on experiments.

sunfishcode commented on lib/frontend/src/frontend.rs:
```diff
@@ -571,10 +571,57 @@ impl<'a> FunctionBuilder<'a> {
+            }
+
+            let flags = MemFlags::new();
+            flags.aligned();
```
This should be `set_aligned()`.

sunfishcode commented on lib/frontend/src/frontend.rs:
```diff
@@ -571,10 +571,57 @@ impl<'a> FunctionBuilder<'a> {
+
+            for i in 0..load_and_store_amount {
+                let offset = (access_size * i) as i32;
+                let value = self.ins().load(types::I8, flags, src, offset);
```
The type for the load here should be an integer type the size of `access_size`. You can use `Type::int` to get such a type.

sunfishcode commented on lib/frontend/src/frontend.rs:
```diff
@@ -591,14 +638,37 @@ impl<'a> FunctionBuilder<'a> {
+    pub fn emit_small_memset(&mut self,
+                             isa: &TargetIsa,
+                             buffer: Value,
+                             ch: Value,
```
I don't think I mentioned this yet, but for the small memset function, we actually want the `ch` to be passed in as a constant too, since the main use for this function is just zeroing out memory.

Also, we want this function to take an alignment parameter similar to memcpy for its destination pointer, and we can compute an access size from it in a similar way. And in that case we'll need to convert `ch` from a single byte to an access-size-sized integer in which every byte is `ch` :).

Aaronepower commented:
@sunfishcode Alright, I've implemented those changes, let me know if there is anything else!

Aaronepower commented:
@sunfishcode I've implemented that feedback.  I also changed how the greatest divisible power of two logic based on this [stackoverflow answer](https://stackoverflow.com/questions/1551775/calculating-highest-power-of-2-that-evenly-divides-a-number-in-c), which doesn't require a starting number so I added an assert so that the alignments are always at lest smaller, let me know if that's the right thing to do or if it's completely off the mark.

sunfishcode commented:
Cool, this looks really good! The CI is red due to formatting diffs; would you mind running rustfmt on this PR?

Aaronepower commented:
@sunfishcode Done!

sunfishcode commented:
Thanks!
